prov/util/src/util_attr.c

```diff
@@ -721,11 +722,52 @@ int ofi_check_tx_attr(const struct fi_provider *prov,
 	return 0;
 }
 
-int ofi_check_info(const struct util_prov *util_prov, uint32_t api_version,
-		   const struct fi_info *user_info)
+/* if there are multiple fi_info in the provider */
+int ofi_check_info_ex(const struct util_prov *util_prov, uint32_t api_version,
+		      const struct fi_info *user_info,
+		      struct fi_info **ret_info)
 {
 	const struct fi_info *prov_info = util_prov->info;
 	const struct fi_provider *prov = util_prov->prov;
+	struct fi_info *info = NULL;
+	int ret, first = 1, res = -FI_ENODATA;
+
+	for ( ; prov_info; prov_info = prov_info->next) {
+		ret = ofi_check_info(util_prov, prov_info,
+				     api_version, user_info);
+	    	if (ret)
+			continue;
+		if (ret_info) {
+			info = fi_dupinfo(prov_info);
+			if (first) {
+				if (!info)
+					goto first;
+				*ret_info = info;
+				first = 0;
+			}
+			if (!info)
+				goto other;
+			info = info->next;
+		}
+		/* found one valid fi_info at least */
+		res = FI_SUCCESS;
+	}
+
+	return res;
+other:
+	fi_freeinfo(*ret_info);
+first:
+	FI_INFO(prov, FI_LOG_CORE,
+		"cannot copy info\n");
+	return -FI_ENOMEM;
+}
```

Comment:
I believe the verbs provider's check functions have some differences w.r.t utility check functions. They have to be addressed first before they can be replaced with util functions. That could be done separately.,

FWIW, verb's provider's fi_ibv_get_matching_info does something similar to what ofi_check_info_ex does.